### PR TITLE
 Check external dependencies versioned symbols

### LINKED ISSUE #146: Repair can produce an invalid manylinux1 wheel
I'm experimenting with building manylinux2010 wheels for pygame, to test the toolchain. `auditwheel repair` (with no `--plat`) argument tags the wheel `manylinux1`. But then re-running auditwheel on that wheel, it is not valid for manylinux1:

- `auditwheel show` says "This constrains the platform tag to "manylinux2010_x86_64". In order
to achieve a more compatible tag, you would need..."
- `auditwheel repair` on the already-repaired wheel says:

> cannot repair "wheelhouse/pygame-2.0.0.dev0-cp37-cp37m-manylinux1_x86_64.whl" to "manylinux1_x86_64" ABI because of the presence of too-recent versioned symbols. You'll need to compile the wheel on an older toolchain.

I understand that it's necessary to specify the tag manually at the command line, but it shouldn't successfully apply the `manylinux1` tag if the wheel is not `manylinux1` compatible.

Educated guesswork: I think it checks the validity of the tag based on the files already in the wheel. These are presumably valid for manylinux1. But then it grafts some libraries into the wheel, and these make it invalid.

### PR DESCRIPTION
**Pull request by mayeut**
This should fix #146 
Uses a different mechanism compared to #148, it checks before producing the wheel.
It checks external dependencies versioned symbols when first analyzing the wheel ABI.

### CODE REVIEW

File: auditwheel/wheel_abi.py [hunk 1]
```diff
@@ -112,7 +113,38 @@ def analyze_wheel_abi(wheel_fn: str):
     log.debug('external reference info')
     log.debug(json.dumps(external_refs, indent=4))
 
+    # get versioned symbols from external_refs
+    # since linux tag (priority 0) has no white-list, do not analyze it
+    external_ref_no_linux = dict(set([(v, k) for e in external_refs.values() if e['priority'] != 0 for k, v in e['libs'].items() if v is not None]))
+    external_refs_symbol_policy = {}
+    log.debug(external_ref_no_linux)
+    for path, elf in elf_file_filter([path for path in iter(external_ref_no_linux)]):
```
Comment (takluyver) [inline auditwheel/wheel_abi.py]: `[path for path in iter(...)]` can be spelled `list(...)`. Or `.keys()` would work, I think.

File: tests/test_manylinux.py [hunk 2]
```diff
@@ -225,6 +225,46 @@ def test_build_wheel_with_binary_executable(docker_container):
...
+
+    # Repair the wheel using the appropriate manylinux container
+    repair_command = (
+        'auditwheel -v repair --plat manylinux1_x86_64 -w /io /io/{orig_wheel}'
+    ).format(policy=policy, orig_wheel=orig_wheel)
+    try:
+        docker_exec(manylinux_id, ['bash', '-c', 'LD_LIBRARY_PATH=/auditwheel_src/tests/testdependencies ' + repair_command])
+        assert policy == 'manylinux1'
+    except CalledProcessError:
+        assert policy != 'manylinux1'
```
Comment (takluyver) [inline tests/test_manylinux.py]: I follow how the test works here, but the logic is kind of inside out. We're really testing that it succeeds when built on the manylinux1 image but not when built on the manylinux2010 image - because then it uses a newer symbol and isn't manylinux1 compatible.

But what we actually assert is that if repairing works, then it must be a manylinux1 image in use, and if it fails, it must *not* be. The dependent and independent variables are swapped around.

That's OK, and I can't immediately see a neat way to write the test without doing that. But let's have an explanatory comment, and maybe a helpful message if the assertions fail.

Comment (mayeut) [reply tests/test_manylinux.py]: > But let's have an explanatory comment

I realize now that the comments does not reflect what's done...
I will have a hard time explaining what's done with decent english but let's try. If you have a better suggestion, please do tell. If you think that's good enough, I'll update the commit.
```
# Always try to repair the wheel targeting manylinux1 policy

# with_dependency == 0
# The python module has no dependencies that should be grafted-in and
# uses versioned symbols not available on policies pre-dating the policy matching the image being tested.

# with_dependency == 1
# The python module itself does not use versioned symbols but has a dependency that should be grafted-in that
# uses versioned symbols not available on policies pre-dating the policy matching the image being tested.

# Since manylinux1 has no pre-dating policy, it's the only image/policy that should be able to repair those wheels.
```

> The dependent and independent variables are swapped around.

Maybe something like that would work:
```
repair_command = (
	        'LD_LIBRARY_PATH=/auditwheel_src/tests/testdependencies auditwheel -v repair --plat manylinux1_x86_64 -w /io /io/{orig_wheel}'
	    ).format(policy=policy, orig_wheel=orig_wheel)
if policy != 'manylinux1':
    with self.assertRaises(CalledProcessError, msg='Repairing this wheel targeting manylinux1  shall fail on {policy} image'.format(policy=policy)):
        docker_exec(manylinux_id, ['bash', '-c', repair_command])
    return

# go-on checking all works properly on manylinux1 image
docker_exec(manylinux_id, ['bash', '-c', repair_command])
...
```

File: tests/test_manylinux.py [hunk 3]
```diff
@@ -225,6 +225,46 @@ def test_build_wheel_with_binary_executable(docker_container):
...
+    assert len(filenames) == 2
+    repaired_wheels = [fn for fn in filenames if policy in fn]
+    # Wheel picks up newer symbols when built in manylinux2010
+    expected_wheel_name = 'testdependencies-0.0.1-cp35-cp35m-%s_x86_64.whl' % policy
+    assert repaired_wheels == [expected_wheel_name]
+    repaired_wheel = repaired_wheels[0]
+    output = docker_exec(manylinux_id, 'auditwheel show /io/' + repaired_wheel)
+    assert (
+        'testdependencies-0.0.1-cp35-cp35m-{policy}_x86_64.whl is consistent'
+        ' with the following platform tag: "{policy}_x86_64"'
```
[the remark this answers is not in the document]
Comment (mayeut) [reply tests/test_manylinux.py]: yes, good point.

File: tests/test_manylinux.py (hunk 2 above)
Comment (takluyver) [reply tests/test_manylinux.py]: I think the rearrangement you propose is an improvement - it's more verbose, but it makes the intent more clear.

I'm happy to have a go at clarifying any comments I can understand, as I am a native English speaker. :slightly_smiling_face:

Comment (mayeut) [reply tests/test_manylinux.py]: Should be better now. I force-pushed a new commit for tests.

File: tests/test_manylinux.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayeut) [reply tests/test_manylinux.py]: Now done. I force-pushed a new commit for tests.

File: auditwheel/wheel_abi.py [hunk 4]
```diff
@@ -112,7 +113,38 @@ def analyze_wheel_abi(wheel_fn: str):
     log.debug('external reference info')
     log.debug(json.dumps(external_refs, indent=4))
 
+    # get versioned symbols from external_refs
+    # since linux tag (priority 0) has no white-list, do not analyze it
+    external_ref_no_linux = dict(set([(v, k) for e in external_refs.values() if e['priority'] != 0 for k, v in e['libs'].items() if v is not None]))
```
[the remark this answers is not in the document]
Comment (mayeut) [reply auditwheel/wheel_abi.py]: I pushed a new commit that I hope clarify things a bit. I'll squash it in the previous commit if that works for you otherwise, let's iterate a bit more before doing that.

File: auditwheel/wheel_abi.py [hunk 5]
```diff
@@ -112,7 +113,46 @@ def analyze_wheel_abi(wheel_fn: str):
...
+    for e in external_refs.values():  # go through all policies
+        if e['priority'] == 0:
+            continue    # since linux tag (priority 0) has no white-list, do not analyze it
+        for libname, realpath in e['libs'].items():  # go through all referenced libs, retrieving their name and realpath
+            if realpath is None:
+                continue  # we can't analyze the lib if we don't know where it's located
+            if (realpath, libname) not in external_ref_no_linux:
+                external_ref_no_linux.append((realpath, libname))
+    external_ref_no_linux = dict(external_ref_no_linux)  # create a dict using the realpath as a key, libname as a value
+    external_refs_symbol_policy = {}  # create a dict using libname as a key, versioned symbols as a value
```
Comment (takluyver) [inline auditwheel/wheel_abi.py]: A comment like this could more concisely be written: `# {libname: versioned_symbols}`

But it's still a bit hard to follow what's going on. Can we expand a bit more what versioned symbols are?

```python
external_refs_symbol_policy = {}  # {libname: {???: set([???]) }}
```

It might also be useful to give some example values alongside the prototype:

```python
# {'libc.so.6': {???: set([???]) }}
```

File: tests/test_manylinux.py [hunk 6]
```diff
@@ -225,6 +225,66 @@ def test_build_wheel_with_binary_executable(docker_container):
...
+
+    repair_command = (
+        'LD_LIBRARY_PATH=/auditwheel_src/tests/testdependencies auditwheel -v repair --plat manylinux1_x86_64 -w /io /io/{orig_wheel}'
+    ).format(policy=policy, orig_wheel=orig_wheel)
+
+    if policy != 'manylinux1':
+        # we shall fail on non manylinux1 images
+        with pytest.raises(CalledProcessError):
+            docker_exec(manylinux_id, ['bash', '-c', repair_command])
+    return
```
Comment (takluyver) [inline tests/test_manylinux.py]: I think the indentation is wrong here, so the rest of the test will never run.

File: tests/test_manylinux.py [hunk 7]
```diff
@@ -225,6 +225,66 @@ def test_build_wheel_with_binary_executable(docker_container):
...
+def test_build_wheel_with_highest_policy_and_image_dependencies(with_dependency, docker_container):
+    # Always try to repair the wheel targeting manylinux1 policy
+    #
+    # with_dependency == 0
+    #   The python module has no dependencies that should be grafted-in and
+    #   uses versioned symbols not available on policies pre-dating the policy matching the image being tested.
+    # with_dependency == 1
+    #   The python module itself does not use versioned symbols but has a dependency that should be grafted-in that
+    #   uses versioned symbols not available on policies pre-dating the policy matching the image being tested.
+    # Since manylinux1 has no pre-dating policy, it's the only image/policy that should be able to repair those wheels.
```
Comment (takluyver) [inline tests/test_manylinux.py]: This sentence doesn't quite match my understanding of how the test works.

We can repair the wheel built on any image, but only to the target platform corresponding to the image. The test checks that a wheel built on manylinux2010 can be repaired to target manylinux1, and that a wheel built on manylinux1 can be.

File: auditwheel/wheel_abi.py [hunk 8]
```diff
@@ -112,7 +113,46 @@ def analyze_wheel_abi(wheel_fn: str):
...
+    external_ref_no_linux = []
+    for e in external_refs.values():  # go through all policies
+        if e['priority'] == 0:
+            continue    # since linux tag (priority 0) has no white-list, do not analyze it
+        for libname, realpath in e['libs'].items():  # go through all referenced libs, retrieving their name and realpath
+            if realpath is None:
+                continue  # we can't analyze the lib if we don't know where it's located
+            if (realpath, libname) not in external_ref_no_linux:
+                external_ref_no_linux.append((realpath, libname))
+    external_ref_no_linux = dict(external_ref_no_linux)  # create a dict using the realpath as a key, libname as a value
```
[the remark this answers is not in the document]
Comment (mayeut) [reply auditwheel/wheel_abi.py]: nope, I just transformed the one-liner without thinking... will change that.

File: tests/test_manylinux.py (hunk 6 above)
Comment (mayeut) [reply tests/test_manylinux.py]: right, will correct

**Comment by mayeut** [comment]
@takluyver, I pushed 2 commits that should answer your comments

**Comment by takluyver** [comment]
Thanks, the comments really do help me understand what's going on.

There are some rather long lines there now. I'm not picky about any particular column limit, but some of these lines are definitely longer than is comfortably readable. Do you have time to reformat the code a bit? Or I can have a go at this if you prefer.

**Comment by takluyver** [comment]
Now #158 added a linter to the test suite, so we will need to stick to some column limit to satisfy that.

**Comment by mayeut** [comment]
I’ll rebase next week in order to get the linter checks

**Comment by mayeut** [comment]
I've rebased this PR on master, it now passes the lint check but I'm not quite happy with the type description of some variables. I don't know how to improve the readability of those comments, so feel free to comment on these.

File: auditwheel/wheel_abi.py [hunk 9]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
...
+        # go through all libs, retrieving their soname and realpath
+        for libname, realpath in e['libs'].items():
+            # we can't analyze the lib if we don't know where it's located
+            if realpath is None:
+                continue
+            if realpath not in external_ref_no_linux.keys():
+                external_ref_no_linux[realpath] = libname
+    # {libname: {depname: set([symbol_version])}} e.g.
+    # {'external_ref.so.1': {'libc.so.6', set(['GLIBC_2.5','GLIBC_2.12'])}}
+    external_refs_symbol_policy = {}
```
Comment (lkollar) [inline auditwheel/wheel_abi.py]: This should be another function, returning the `external_refs_symbol_policy` dict.

File: auditwheel/wheel_abi.py [hunk 10]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
...
+    for path, elf in elf_file_filter(external_ref_no_linux.keys()):
+        # {depname: set(symbol_version)}, e.g.
+        # {'libc.so.6', set(['GLIBC_2.5','GLIBC_2.12'])}
+        elf_versioned_symbols = defaultdict(lambda: set())
```
Comment (lkollar) [inline auditwheel/wheel_abi.py]: `defaultdict(set)` should work.

File: auditwheel/wheel_abi.py [hunk 11]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
...
+        for key, value in elf_find_versioned_symbols(elf):
+            log.debug('path %s, key %s, value %s', path, key, value)
+            elf_versioned_symbols[key].add(value)
+        external_refs_symbol_policy[external_ref_no_linux[path]] = \
```
Comment (lkollar) [inline auditwheel/wheel_abi.py]: This would be more readable as two separate assignments and it would probably fit without a break.

```
libname = external_ref_no_linux[path]
external_refs_symbol_policy[libname] = elf_versioned_symbols
```

File: auditwheel/wheel_abi.py [hunk 12]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
...
+            elf_versioned_symbols
+
     symbol_policy = versioned_symbols_policy(versioned_symbols)
+
+    # since white-list is different per policy,
+    # let's examine versioned_symbol per policy when including external refs
+    # list of tuples of the form (policy_priority, versioned_symbols), e.g.
+    # [(100, {'libc.so.6', set(['GLIBC_2.5'])})]
+    symbol_policies = []
```
Comment (lkollar) [inline auditwheel/wheel_abi.py]: This should be a separate function too, returning `symbol_policies`.

File: auditwheel/wheel_abi.py [hunk 13]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
     log.debug('external reference info')
     log.debug(json.dumps(external_refs, indent=4))
 
+    # get versioned symbols from external_refs
+    # {realpath: libname} e.g.
+    # {'/path/to/external_ref.so.1.2.3': 'external_ref.so.1'}
+    external_ref_no_linux = {}
+    # go through all policies
+    for e in external_refs.values():
```
Comment (lkollar) [inline auditwheel/wheel_abi.py]: I would much prefer proper variable names in the loops instead of single letter names. I'm surprised the linter doesn't complain about this.

File: tests/test_manylinux.py [hunk 14]
```diff
@@ -225,6 +226,63 @@ def test_build_wheel_with_binary_executable(docker_container):
         python_id, ['python', '-c', 'from testpackage import runit; print(runit(1.5))']).strip()
     assert output.strip() == '2.25'
 
+@pytest.mark.parametrize('with_dependency', ['0', '1'])
+def test_build_wheel_with_highest_policy_and_image_dependencies(with_dependency, docker_container):
```
Comment (lkollar) [inline tests/test_manylinux.py]: Can you try to make the line lengths shorter to make it more consistent with the rest of the file? We don't run the linter on tests at the moment but the rest of the tests are mostly under 80 characters.

Comment (lkollar) [inline tests/test_manylinux.py]: I really like this test and the native extension.

File: tests/test_manylinux.py [hunk 15]
```diff
@@ -225,6 +226,63 @@ def test_build_wheel_with_binary_executable(docker_container):
...
+    # with_dependency == 0
+    #   The python module has no dependencies that should be grafted-in and
+    #   uses versioned symbols not available on policies pre-dating the policy matching the image being tested.
+    # with_dependency == 1
+    #   The python module itself does not use versioned symbols but has a dependency that should be grafted-in that
+    #   uses versioned symbols not available on policies pre-dating the policy matching the image being tested.
+
+    policy, manylinux_id, python_id, io_folder = docker_container
+
+    docker_exec(manylinux_id, ['bash', '-c', 'cd /auditwheel_src/tests/testdependencies && WITH_DEPENDENCY={} python setup.py -v build_ext -f bdist_wheel -d /io'.format(with_dependency)])
```
Comment (lkollar) [inline tests/test_manylinux.py]: Wouldn't `pip wheel` work here? It would be simpler. There is an example in `test_build_wheel_with_binary_executable` for this.

Comment (mayeut) [reply tests/test_manylinux.py]: I can't use `pip wheel` because I need to force rebuilding the extension between tests (otherwise the extension used in a previous test will be reused without the correct `WITH_DEPENDENCY` option)

File: tests/test_manylinux.py (hunk 14 above)
Comment (mayeut) [reply tests/test_manylinux.py]: done

File: auditwheel/wheel_abi.py [hunk 16]
```diff
@@ -116,7 +117,65 @@ def analyze_wheel_abi(wheel_fn: str):
...
+        # linux tag (priority 0) has no white-list, do not analyze it
+        if e['priority'] == 0:
+            continue
+        # go through all libs, retrieving their soname and realpath
```
[the remark this answers is not in the document]
Comment (takluyver) [reply auditwheel/wheel_abi.py]: Sorry, I was the one pushing for more comments because I didn't understand the code, and it's possible that I pushed a bit too much.

**Comment by lkollar** [comment]
@auvipy I was going to review the changes. Why did you merge this? In the future could you wait until the person who requested changes approves the final version?

**Comment by auvipy** [comment]
OK sure.
